Information/CenterOfMass.FCMacro

```diff
@@ -1,891 +1,966 @@
+        # Only nested App:Part are considered at this place.
+        groupObjs = ['App::DocumentObjectGroup', 'App::GeometryPython']
+        i_ = 0
+        while i_ < len(_sel):
+            if hasattr(_sel[i_], 'Group') and _sel[i_].TypeId in groupObjs:
+                _sel.extend(_sel[i_].Group)
+                del _sel[i_]
+            elif hasattr(_sel[i_], 'Group') and _sel[i_].TypeId == 'App::Part':
+                for gp in _sel[i_].Group:
+                    if gp.TypeId == 'App::Part':
+                        _sel.append(gp)
+                i_ += 1    # where to look at the next run
+            else:
+                i_ += 1
+
+        # create valid selection list
+        vsel = []
+        for s_ in _sel:
+            if hasattr(s_, 'Shape') and s_.Shape.Volume:
+                if s_.TypeId == 'App::Part':
+                    # because contains bodies
+                    csList = s_.Shape.childShapes(False,False)    # ignore placement of parent
+                    for cs in range(len(csList)):
+                        # get first match from OutList (contains different object types)
+                        for ot in s_.OutList:
+                            if hasattr(ot, 'Shape') and ot.Shape.isEqual(csList[cs]):
+                                # nested App:Part and App:Link in App:Part not match here
+                                vsel.append(ot)
+                                break
+                    # App:Link in App:Part
+                    for ot in s_.OutList:
+                        if ot.TypeId == 'App::Link':
+                            vsel.append(ot)
+                elif s_.TypeId == 'Part::FeaturePython' and hasattr(s_, 'IfcType'):
+                    # e.g. Arch Wall/Structure, because can contain childs
+                    vsel.append(s_)
+                    for it in s_.InList[1:]:
+                        # include childs (Windows etc.), parent is first InList
+                        if it.TypeId == 'Part::FeaturePython' and it.Shape.Volume:
+                            vsel.append(it)
+                else:
+                    vsel.append(s_)
+            elif hasattr(s_, 'Mesh') and s_.Mesh.Volume:
+                vsel.append(s_)
 
-    def on_pushButton_EditMat(self):
-        MaterialEditor.openEditor()
-        self.loadMaterials()
+        # sort valid selection list or match up with tree view
+        if sort_selection:
+            vsel.sort(key=sortSelectionByLabel)
+        else:
+            tree = FreeCADGui.getMainWindow().findChild(QtGui.QTreeWidget)
+            iterator = QtGui.QTreeWidgetItemIterator(tree, QtGui.QTreeWidgetItemIterator.Editable)
+            # Syntax: [ EXP for x in seq if COND ]
+            self.tree_list = [i_.value().text(0) for i_ in list(iterator)]
+            vsel.sort(key=sortSelectionByTree)
+
+        # create valid objects list (e.g. shapes)
+        objs = []
+        for i_ in range(len(vsel)):
```

Comment:
Why not `for s_ in vsel:`?

Reply:
I need to write back to `vsel` in the last line of the loop